src/coder/decode.js

```diff
@@ -0,0 +1,61 @@
+'use strict'
+
+const varint = require('varint')
+const BufferList = require('bl')
+
+// Decode a chunk and yield an _array_ of decoded messages
+module.exports = source => (async function * decode () {
+  const decoder = new Decoder()
+  for await (const chunk of source) {
+    const msgs = decoder.write(chunk)
+    if (msgs.length) yield msgs
+  }
+})()
+
+class Decoder {
+  constructor () {
+    this._buffer = new BufferList()
+    // optimisation to allow varint to take a bl (well a proxy to)
+    this._bufferProxy = new Proxy({}, {
+      get: (_, prop) => prop[0] === 'l' ? this._buffer[prop] : this._buffer.get(parseInt(prop))
+    })
+    this._headerInfo = null
+  }
+
+  write (chunk) {
+    if (!chunk || !chunk.length) return []
+
+    this._buffer.append(chunk)
+    const msgs = []
+
+    while (true) {
+      if (!this._headerInfo) {
+        try {
+          this._headerInfo = this._decodeHeader(this._bufferProxy)
+        } catch (_) {
+          break // not enough data yet...probably
```

Comment:
Can you elaborate a bit more? The "probably" is not really comforting :)

Reply:
Updated the comment for this. We should only hit the break in 2 instances. 
1. We haven't gotten enough data to parse a header (most common).
2. The data we have can't be parsed, likely due to an invalid varint from bad data.